src/NServiceBus.Core/Pipeline/Outgoing/SerializeMessageConnector.cs

```diff
@@ -52,10 +53,16 @@ string SerializeEnclosedMessageTypes(Type messageType)
         {
             var metadata = messageMetadataRegistry.GetMessageMetadata(messageType);
 
-            var assemblyQualifiedNames = new HashSet<string>();
+            var assemblyQualifiedNames = new List<string>();
             foreach (var type in metadata.MessageHierarchy)
             {
-                assemblyQualifiedNames.Add(type.AssemblyQualifiedName);
+                var typeAssemblyQualifiedName = type.AssemblyQualifiedName;
+                if (assemblyQualifiedNames.Contains(typeAssemblyQualifiedName))
```

Comment:
Since we're likely dealing with relatively low numbers of items in `metadata.MessageHierarchy`, I suppose adding an O(N) call here isn't too big of a deal.

Reply:
The alternative would be to use `Distinct` which works with an internal set for the deduplication check, but since the expected amount of entries in these collections should be very small (< 5), I think this shouldn't be any issue.

Reply:
For reference the PR which introduced this https://github.com/Particular/NServiceBus/pull/4009

Reply:
How about doing a distinct in https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Unicast/Messages/MessageMetadataRegistry.cs#L113 or would that break other things?